File: 프로그래머스/lv3/42895. N으로 표현/N으로 표현.py
from collections import deque


def solution(N, number):
    dp = [set() for _ in range(9)]

    for i in range(1, 9):
        target = dp[i]
        target.add(int(str(N) * i))
        
        for j in range(1, i):
            for k in dp[j]:
                for l in dp[i - j]:
                    if k < number:
                        target.add(k + l)
                        target.add(k * l)
                    if k > 0:
                        target.add(k - l)
                    if k != 0 and l != 0:
                        target.add(k // l)
        
        if number in target:
            return i
        
    return -1

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.java
import java.util.*;

class Solution {
    
    public int solution(int N, int number) {
        if (N == number) {
            return 1;
        }

        List<Set<Integer>> dp = new ArrayList<>();
        for (int i = 0; i <= 8; i++) {
            dp.add(new HashSet<>());
        }

        for (int i = 1; i <= 8; i++) {
            Set<Integer> target = dp.get(i);
            target.add(Integer.parseInt(String.valueOf(N).repeat(i)));

            for (int j = 1; j < i; j++) {
                for (int k : dp.get(j)) {
                    for (int l : dp.get(i - j)) {
                        target.add(k + l);
                        if (k - l > 0) {
                            target.add(k - l);
                        }
                        if (k * l <= number) {
                            target.add(k * l);
                        }
                        if (k != 0 && l != 0) {
                            target.add(k / l);
                        }
                    }
                }
            }

            if (target.contains(number)) {
                return i;
            }
        }

        return -1;
    }
}